### PR TITLE
feat: Token transfers list API endpoint

### PR DESCRIPTION
**Pull request by vbaranov**
Resolves https://github.com/blockscout/blockscout/issues/10717

## Motivation

Development of API v2 endpoint for the list of token transfers. 

## Changelog

GET `api/v2/token-transfers` API v2 endpoint is added.
It supports keyset pagination in the form:
`?block_number=6715325&index=24&items_count=50` which is taken from `next_page_params` param of the response for the 1st page:
```
    "next_page_params": {
        "block_number": 6715322,
        "index": 254,
        "items_count": 100
    }
```

Sample response:
```
{
    "items": [
{
            "block_hash": "0x36606ce63df214e78ca6fc391997b3f608934039caac6415e8aed0ff80b13def",
            "block_number": "6788861",
            "from": {
                "ens_domain_name": null,
                "hash": "0x0000000000000000000000000000000000000000",
                "implementations": [],
                "is_contract": false,
                "is_scam": true,
                "is_verified": false,
                "metadata": {
                    "tags": [
                        {
                            "meta": {},
                            "name": "METAMASK USER",
                            "ordinal": 0,
                            "slug": "metamask-user",
                            "tagType": "generic"
                        },
                        {
                            "meta": {},
                            "name": "Null: 0x000...000",
                            "ordinal": 0,
                            "slug": "null-0x000000",
                            "tagType": "name"
                        },
                        {
                            "meta": {},
                            "name": "IMTOKEN USER",
                            "ordinal": 0,
                            "slug": "imtoken-user",
                            "tagType": "generic"
                        },
                        {
                            "meta": {},
                            "name": "MINER",
                            "ordinal": 0,
                            "slug": "miner",
                            "tagType": "generic"
                        },
                        {
                            "meta": {},
                            "name": "Null: 0x000...000",
                            "ordinal": 0,
                            "slug": "null-0x000000",
                            "tagType": "name"
                        }
                    ]
                },
                "name": null,
                "private_tags": [],
                "proxy_type": null,
                "public_tags": [],
                "watchlist_names": []
            },
            "log_index": "61",
            "method": "0x00fa478c",
            "timestamp": "2024-09-30T16:33:00.000000Z",
            "to": {
                "ens_domain_name": null,
                "hash": "0x0985b6C72e73b27FF1a768FDA30394ce14Cc3b6D",
                "implementations": [],
                "is_contract": false,
                "is_scam": true,
                "is_verified": false,
                "metadata": null,
                "name": null,
                "private_tags": [],
                "proxy_type": null,
                "public_tags": [],
                "watchlist_names": []
            },
            "token": {
                "address": "0xF7E71ed8ed39031F57e71381A070216c8dA345DC",
                "circulating_market_cap": null,
                "decimals": null,
                "exchange_rate": null,
                "holders": "3",
                "icon_url": null,
                "name": null,
                "symbol": null,
                "total_supply": null,
                "type": "ERC-1155",
                "volume_24h": null
            },
            "total": {
                "decimals": null,
                "token_id": "3",
                "value": "1"
            },
            "tx_hash": "0x7300b509c1c727a631789c1c1fbce291cbf293783514fa3a74966faa5012ce42",
            "type": "token_minting"
        },
        {
            "block_hash": "0xba6656e4d7726229b5fcfcccd29d255176dc5251ff9b47d72083867a29755ef9",
            "block_number": "6788851",
            "from": {
                "ens_domain_name": null,
                "hash": "0x0000000000000000000000000000000000000000",
                "implementations": [],
                "is_contract": false,
                "is_scam": true,
                "is_verified": false,
                "metadata": {
                    "tags": [
                        {
                            "meta": {},
                            "name": "METAMASK USER",
                            "ordinal": 0,
                            "slug": "metamask-user",
                            "tagType": "generic"
                        },
                        {
                            "meta": {},
                            "name": "Null: 0x000...000",
                            "ordinal": 0,
                            "slug": "null-0x000000",
                            "tagType": "name"
                        },
                        {
                            "meta": {},
                            "name": "IMTOKEN USER",
                            "ordinal": 0,
                            "slug": "imtoken-user",
                            "tagType": "generic"
                        },
                        {
                            "meta": {},
                            "name": "MINER",
                            "ordinal": 0,
                            "slug": "miner",
                            "tagType": "generic"
                        },
                        {
                            "meta": {},
                            "name": "Null: 0x000...000",
                            "ordinal": 0,
                            "slug": "null-0x000000",
                            "tagType": "name"
                        }
                    ]
                },
                "name": null,
                "private_tags": [],
                "proxy_type": null,
                "public_tags": [],
                "watchlist_names": []
            },
            "log_index": "21",
            "method": "0x00fa478c",
            "timestamp": "2024-09-30T16:30:36.000000Z",
            "to": {
                "ens_domain_name": null,
                "hash": "0x0985b6C72e73b27FF1a768FDA30394ce14Cc3b6D",
                "implementations": [],
                "is_contract": false,
                "is_scam": true,
                "is_verified": false,
                "metadata": null,
                "name": null,
                "private_tags": [],
                "proxy_type": null,
                "public_tags": [],
                "watchlist_names": []
            },
            "token": {
                "address": "0xF7E71ed8ed39031F57e71381A070216c8dA345DC",
                "circulating_market_cap": null,
                "decimals": null,
                "exchange_rate": null,
                "holders": "3",
                "icon_url": null,
                "name": null,
                "symbol": null,
                "total_supply": null,
                "type": "ERC-1155",
                "volume_24h": null
            },
            "total": {
                "decimals": null,
                "token_id": "2",
                "value": "1"
            },
            "tx_hash": "0xb2e597ed96f051509f2156bd5f0dbe27a2635f20b49eb008dc4c4990825bf3a5",
            "type": "token_minting"
        },
        ...
    ],
    "next_page_params": {
        "block_number": 6715322,
        "index": 254,
        "items_count": 100
    }
}
```

## Checklist for your Pull Request (PR)

- [x] If I added new functionality, I added tests covering it.
- [ ] If I fixed a bug, I added a regression test to prevent the bug from silently reappearing again.
- [x] I checked whether I should update the docs and did so by submitting a PR to [docs repository](https://github.com/blockscout/docs).
- [x] If I added/changed/removed ENV var, I submitted a PR to [docs repository](https://github.com/blockscout/docs) to update the list of [env vars](https://github.com/blockscout/docs/blob/master/for-developers/information-and-settings/env-variables.md) and I updated the version to `master` in the Version column. If I removed variable, I added it to [Deprecated ENV Variables](https://github.com/blockscout/docs/blob/master/for-developers/information-and-settings/env-variables/deprecated-env-variables/README.md) page. After merging docs PR, changes will be reflected in these [pages](https://docs.blockscout.com/for-developers/information-and-settings/env-variables).
- [x] If I added new DB indices, I checked, that they are not redundant, with PGHero or other tools.
- [x] If I added/removed chain type, I modified the Github CI matrix and PR labels accordingly.

### CODE REVIEW

File: apps/block_scout_web/lib/block_scout_web/controllers/api/v2/token_transfer_controller.ex [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+      |> Enum.zip(decoded_transactions)
+      |> Enum.into(%{}, fn {%{hash: hash}, decoded_input} -> {hash, decoded_input} end)
+
+    next_page_params = next_page |> next_page_params(token_transfers, delete_parameters_from_next_page_params(params))
+
+    conn
+    |> put_status(200)
+    |> put_view(TokenTransferView)
+    |> render(:token_transfers, %{
+      token_transfers: token_transfers,
```
Comment (nikitosing) [inline apps/block_scout_web/lib/block_scout_web/controllers/api/v2/token_transfer_controller.ex:65]: let's add  `|> maybe_preload_ens() |> maybe_preload_metadata()`?

File: apps/block_scout_web/lib/block_scout_web/views/api/v2/token_transfer_view.ex [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+  def render("token_transfers.json", %{
+        token_transfers: token_transfers,
+        decoded_transactions_map: decoded_transactions_map,
+        next_page_params: next_page_params
+      }) do
+    %{
+      "items" =>
+        Enum.map(
+          token_transfers,
+          &render("token_transfer.json", %{
```
Comment (nikitosing) [inline apps/block_scout_web/lib/block_scout_web/views/api/v2/token_transfer_view.ex:41]: Why didn't you use `BlockScoutWeb.API.V2.TransactionView.prepare_token_transfer/3`

File: apps/block_scout_web/lib/block_scout_web/views/api/v2/token_transfer_view.ex [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+  def render("token_transfer.json", %{
+        token_transfer: token_transfer,
+        decoded_transaction_input: decoded_transaction_input
+      }) do
+    %{
+      "transaction_hash" => token_transfer.transaction_hash,
+      "block_number" => token_transfer.block_number,
+      "log_index" => token_transfer.log_index,
+      "block_timestamp" => token_transfer.transaction.block_timestamp,
+      "amounts_or_token_ids" => token_transfer.amounts || [token_transfer.amount || 1],
```
Comment (nikitosing) [inline apps/block_scout_web/lib/block_scout_web/views/api/v2/token_transfer_view.ex:20]: seems that `token_ids` is missing

File: apps/block_scout_web/lib/block_scout_web/controllers/api/v2/token_transfer_controller.ex [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+    decoded_transactions_map =
+      transactions
+      |> Enum.zip(decoded_transactions)
+      |> Enum.into(%{}, fn {%{hash: hash}, decoded_input} -> {hash, decoded_input} end)
+
+    next_page_params = next_page |> next_page_params(token_transfers, delete_parameters_from_next_page_params(params))
+
+    conn
+    |> put_status(200)
+    |> put_view(TokenTransferView)
```
Comment (nikitosing) [inline apps/block_scout_web/lib/block_scout_web/controllers/api/v2/token_transfer_controller.ex:63]: I think we can remove this line

Comment (vbaranov) [reply apps/block_scout_web/lib/block_scout_web/controllers/api/v2/token_transfer_controller.ex]: Yep, removed in https://github.com/blockscout/blockscout/pull/10801/commits/45cc4c64e4c8caa554781614cc89457977d7ffb7.

File: apps/block_scout_web/lib/block_scout_web/views/api/v2/token_transfer_view.ex (hunk 3 above)
Comment (vbaranov) [reply apps/block_scout_web/lib/block_scout_web/views/api/v2/token_transfer_view.ex]: Good catch, added in https://github.com/blockscout/blockscout/pull/10801/commits/45cc4c64e4c8caa554781614cc89457977d7ffb7.

File: apps/block_scout_web/lib/block_scout_web/controllers/api/v2/token_transfer_controller.ex (hunk 1 above)
Comment (vbaranov) [reply apps/block_scout_web/lib/block_scout_web/controllers/api/v2/token_transfer_controller.ex]: Added in https://github.com/blockscout/blockscout/pull/10801/commits/45cc4c64e4c8caa554781614cc89457977d7ffb7

File: apps/block_scout_web/lib/block_scout_web/views/api/v2/token_transfer_view.ex (hunk 2 above)
Comment (vbaranov) [reply apps/block_scout_web/lib/block_scout_web/views/api/v2/token_transfer_view.ex:41]: Honestly, I don't see pros to refactor code for usage of that function.

Comment (vbaranov) [reply apps/block_scout_web/lib/block_scout_web/views/api/v2/token_transfer_view.ex:32]: I re-thought and re-used `prepare_token_transfer/3` function, preliminary moving it to `BlockScoutWeb.API.V2.TokenTransferView` module. It makes sense to use that function in order to equalize responses from API endpoints which return token transfers.

File: apps/block_scout_web/lib/block_scout_web/controllers/api/v2/token_transfer_controller.ex [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+      |> Enum.uniq()
+
+    {decoded_transactions, _, _} = Transaction.decode_transactions(transactions, true, @api_true)
+
+    decoded_transactions_map =
+      transactions
+      |> Enum.zip(decoded_transactions)
+      |> Enum.into(%{}, fn {%{hash: hash}, decoded_input} -> {hash, decoded_input} end)
+
+    next_page_params = next_page |> next_page_params(token_transfers, delete_parameters_from_next_page_params(params))
```
Comment (nikitosing) [inline apps/block_scout_web/lib/block_scout_web/controllers/api/v2/token_transfer_controller.ex:68]: Here should be token_transfers_next_page_params function call

**Review by nikitosing — APPROVED** [review]

**Pull request merged**